# Transferring Air Quality Data from Device manufacturer sites to Influxdb
For flexibility and easy of data manipulation, we transfer the data to influx db.
This will also enable us to have the data downloaded and stored in one place - giving us ownership incase our subscription with the device manufacturer expires.

We shall make API calls to the device urls and send the data to the influxdb database. Since different manufacturers have different procedures for API calls, we do upload the data in different batches based on the manufacturer; Quantaq, Clarity, Airgradient

## 0. Set up

### Prepare tools for the project
First install the packages and modules necessary to implement the steps of the project.

The documentation for the influxdb_client_3 library can be found [here](https://docs.influxdata.com/influxdb3/cloud-dedicated/reference/client-libraries/v3/python/)

In [64]:
# Install operational packages to help in
# organizing and querrying the data
import pandas as pd
import numpy as np
import requests
import json
from datetime import datetime
import re  # Allows to easily remove text using regular expressions
from time import sleep  # Allows pauses during api clalls

# Modules to help interact with influxdb
import os, time
from influxdb_client_3 import InfluxDBClient3, Point

# Modules to modify and manipulate time and dates
from datetime import date, timedelta

# Library to help with cleaning column names
import janitor

# Library to help with easily getting
# csv files stored in one folder at once
import glob

Store important athaurization keys to access different databases

In [65]:
#Load the api key and device serial numbers from the envrionment variable 
%reload_ext dotenv
%dotenv
QUANTAQ_APIKEY = os.environ.get("QUANTAQ_APIKEY")
QUANT001_SN = os.environ.get("QUANT001_SN")
QUANT002_SN = os.environ.get("QUANT002_SN")
CLARITY_APIKEY = os.environ.get("CLARITY_APIKEY")
CLARITY_ORGID = os.environ.get("CLARITY_ORGID")
AIRGRADIENT_APIKEY = os.environ.get("AIRGRADIENT_APIKEY")
EPA_KEY = os.environ.get("EPA_KEY")
EPA_ID = os.environ.get("EPA_ID")
AIRNOW_ID = os.environ.get('AIRNOW_ID')
AIRNOW_KEY = os.environ.get('AIRNOW_KEY')
INFLUXDB_TOKEN = os.environ.get('INFLUXDB_TOKEN')



Create a magic function to help in skipping cells so that they run only once. Some cells especially those that involve uplaoding data to the database, can be skipped so that the data is not duplicated when the cells are are run multiple times

In [66]:
from IPython.core.magic import register_cell_magic

@register_cell_magic
def skip(line, cell):
    return

### Set up influxdb
Prepare the database to receive data from the device manufacturer urls 

In [71]:
#add this to load and upadte the token file
%reload_ext dotenv
%dotenv

#State the possible databases for the collected data
databases_list = ['airgradient', 'clarity', 'quantaq', 'zz-airnow']

#Create a dictionary to collect the different client objects
clients_dict = dict()
for db in databases_list:

    #Define some of the arguments required 
    #to write the data to the database

    token = os.environ.get("INFLUXDB_TOKEN")
    org         =   "Air quality - Harrisburg and Philadelphia"
    host        =   "https://us-east-1-1.aws.cloud2.influxdata.com"
    database    =   db

    clients_dict[db] = InfluxDBClient3(host=host, token=token, org=org, database=database)


For automation, Write a function that will later be used to easily upload the data into an influxdb database

In [72]:
def upload_to_influxdb(data_frame_to_save, measurement_name, 
                       index_field, client_name, tag_cols_list):
    """
    A function that takes the name of the measurement(table) 
    and one field name to serve as the index and wites them into
    a defined database on influxdb

    Args:
        data_frame_to_save(df)  : The dataframe to be sent to the database
        measurement_name(string) : The measurement or table name to store the dataset
        index_field(string)     : A timefield from the measurement to be used as an index
                                This also serves as the timeseries reference column
        client_name(object)     : The influxdb object with a defined database 
        tag_cols_list (list)    : A list of column names to act as tags

    Return : 
            Prints that it completed the task
    """ 

    #Make the index field to be the index so that it acts as the timestamp
    data_frame_to_save_index = data_frame_to_save.set_index(index_field)
    
    #Write to the database and the let user know that it is written
    client_name.write(record=data_frame_to_save_index, 
                      data_frame_measurement_name=measurement_name,
                      data_frame_tag_columns=tag_cols_list)
    
    print("Done! Check your influxdb to confirm!")


Since it is not straightforward to delete tables in influxdb; Set the table names here so that when there is a need to delete the table; we can delete the entire database instead and update the table names below

In [73]:
#Measure names defined 
quant_min_table = "pilot-min-quantaq"
quant_hour_table = "pilot-hour-quantaq"
quant_min_colloc_table = "collocation-min-quantaq"
quant_hour_colloc_table = "collocation-hour-quantaq"

clarity_min_table = "pilot-min-clarity"
clarity_hour_table = "pilot-hour-clarity"
clarity_min_colloc_table = "collocation-min-clarity"
clarity_hour_colloc_table = "collocation-hour-clarity"


airgradient_min_table = "pilot-min-airgradient"
airgradient_hour_table = "pilot-hour-airgradient"
airgradient_min_colloc_table = "collocation-min-airgradient"
airgradient_hour_colloc_table = "collocation-hour-airgradient"
airgradient_hour_gases_table = "collocation-hour-gases-ag-clean"

airnow_hour_table = "pilot-hour-airnow"
airnow_hour_colloc_table = "collocation-hour-airnow"
dep_min_colloc_table = "collocation-min-dep"


## 1. Quantaq

The guidelines to make API calls to quantaq can be found [here](https://docs.quant-aq.com/software-apis-and-libraries/quantaq-cloud-api). For using python to manipulate data; find the guidance under the [documentation for the py-quantaq](https://quant-aq.github.io/py-quantaq/usage.html#list-all-final-data-for-a-device); which is a quantaq python library for data manipulation. 

### 1.1 Quantaq Setup and Automation

Using the py-QuantAQ library to get data from the web and manipulate it 


In [81]:
%%skip
pip install py-quantaq

In [82]:
#Bring in the the quantaq libray and instantiate the client object
from quantaq.utils import to_dataframe
from quantaq import QuantAQAPIClient

client_quantaq = QuantAQAPIClient(api_key=QUANTAQ_APIKEY)

For automation, write a function that gets the data from quantaq website given a time frame. This function can be used to obtain the intial batch of data and subsequently used to update the database

In [83]:
def get_quantaq_data_devices(devices_sn_list, 
                             start_time, 
                             end_time):
    """
    A function that accepts a list of devices and returns the 
    data of different parameters given a start date and end date
    
    Args:
        devices_sn_list(list) : Sns for devices from which to collect data
        start_time: (string)  : Time in the format of 'yyyy-mm-dd'
        end_time: (string)  : Time in the format of 'yyyy-mm-dd'

    Return: 
        quantaq_all_data_df(Dataframe) : Dataframe with the requested data

    """

    #Initialize the dataframe to be used to make a 
    # singular table to save to the influxdb database
    quantaq_all_data_df = []

    #Collect the data all the devices from commissioning date to jan-13-2026
    for device in devices_sn_list:
        for each in pd.date_range(start = start_time, end = end_time):
            quantaq_all_data_df.append(
                to_dataframe(client_quantaq.data.bydate(sn=device, 
                                                        date=str(each.date())))
            )
    quantaq_all_data_df = pd.concat(quantaq_all_data_df)

    #For simplicity remove the model and weather fields 
    columns_to_remove=[]
    for col in quantaq_all_data_df.columns:
        if col[:5]=='model':
            columns_to_remove.append(col) 
        if col[:3]=='met':
            columns_to_remove.append(col) 
    
    #create a new dataframe without the columns
    quant_all_modified = quantaq_all_data_df.copy()

    quant_all_modified.drop(columns_to_remove, 
                            axis=1, 
                            inplace=True)
    #Convert the CO parameters from ppb to ppm (standard epa units)
    quant_all_modified['co'] = quant_all_modified['co'] / 1000

    return quant_all_modified

The pandas resampling function uses the timestamp as the index in the sampling process, making it difficult to have two or more devices in the same sampling dataframe since several records are likely to have shared timestamps.

Let's write a function that divides the data into separate dataframes according to the device serial numbers, resamples each and combines the dataframes again

In [84]:
def resample_quantaq(df, frequency='1h'):
    """
    The resample_quantaq takes a quantaq dataframe and 
    Returns a dataframe resampled at a given frequency

    Args:
        df (object)        :A dataframe of quantaq data 
                            collected at minute level
        frequency (string) :Frequency of sampling/aggregating
                            see the resample function in pandas
    
    Return:
        df_all_sampled (object) :A dataframe with data resampled 
                             at the stated interval
    """

    #First convert the timestamp to datetime to pervent any errors
    df['timestamp'] = pd.to_datetime(df['timestamp'])

    #Define the dictionary to be used in resampling so that we can 
    # get the mean of numerical fields wiothout losing categorical ones
    cols_dict = {'co': 'mean', 'no':'mean', 
             'no2':'mean', 'o3':'mean', 
             'pm1':'mean', 'pm10':'mean', ''
             'pm25':'mean', 'raw_data_id':'first', 
             'sn':'first',  
             'timestamp_local':'first', 'url':'first', ''
             'geo.lat':'first', 'geo.lon':'first'}
    
    #Define a dataframe to combine all the dataframes for each device
    df_all_sampled = []    

    for device_sn in pd.unique(df.sn):
        #get a dataframe for each device separately
        df_device = df[df['sn']==device_sn]

        #Resample the dataframe of the device 
        df_device_sampled = (df_device.resample('1h',
                                               on='timestamp',).
                                               agg(cols_dict).
                                               reset_index())

        #Add it to the overall dataframe to be returned 
        df_all_sampled.append(df_device_sampled)

    df_all_sampled = pd.concat(df_all_sampled)    

    return df_all_sampled  
 

### 1.2 QuantAQ Minute Data

##### 1.2.1 First Batch of the QuantAQ minute data to influxDB

Get the first batch of the quantaq data from the quantaq website and store the file as pickle so as to avoid length periods of downloading next time. To allow a buffer in the time ranges; It is advisable to have the start date atleast a day before the intended range and the end date atleast a day after (beyond the range).

In [85]:
%%skip

quantaq_all_data_df = get_quantaq_data_devices(devices_sn_list=['MOD-X-00959','MOD-X-00958'], 
                                               start_time='2025-10-01',
                                               end_time='2026-01-03')

quantaq_all_data_df.to_pickle('data-from-monitors/quantaq_all_devices_first.pkl')

In [86]:
quant_all_first = pd.read_pickle('data-from-monitors/quantaq_all_devices_first.pkl')

Send the initial quantaq batch to the influxdb database for further querrying and vizualization

In [87]:
%%skip
# Use the upload function to send this datframe to inlflux
upload_to_influxdb(
    data_frame_to_save=quant_all_first,
    measurement_name=quant_min_table,
    index_field="timestamp",
    client_name=clients_dict["quantaq"],
    tag_cols_list=["sn", "geo.lat", "geo.lon"],
)

##### 1.3 Update the Quantaq minute data to influxdb Database

Get the most recent date from the intended quantaq influxdb measurement (table). This date is to be used as the start date in the update process. Then use two days from now as the end date - to allow a room for error.

In [88]:
query_date = """SELECT *
                FROM  'pilot-min-quantaq' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the queery on the influxdb database
table = clients_dict['quantaq'].query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from clarity to influxdb
most_recent_quantaq_df = table.to_pandas()
latest_date_quant = str(most_recent_quantaq_df.time[0])[:10]

#Use two days from now as the end date - for the purposes of buffering
end_date_quant = f'{date.today() + timedelta(days=2)}'

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_quant}')
print(f'start: {end_date_quant}')


start: 2026-06-15
start: 2026-06-18


Use the dates above to get the most recent dataframe from quantaq, then uplaod the data to influxdb.

In [89]:
#%%skip
quantaq_update_min = get_quantaq_data_devices(devices_sn_list=['MOD-X-00959','MOD-X-00958'],
                                             start_time=latest_date_quant,
                                             end_time=end_date_quant)



In [90]:
#%%skip
upload_to_influxdb(data_frame_to_save = quantaq_update_min, 
                   measurement_name=quant_min_table, 
                   index_field='timestamp', 
                   client_name=clients_dict["quantaq"],
                   tag_cols_list=['sn','geo.lat','geo.lon'])

Done! Check your influxdb to confirm!


### 1.3 QuantAQ Hour Data

##### The First Batch QuantAQ Hour

Resample the data at 1 hour interval before sending it to influxdb. Simple use the batch in the minute category

In [91]:
#%%skip
quant_all_first_hr = resample_quantaq(quant_all_first)

Now upload te resampled data to influxdb

In [92]:
%%skip
# Use the upload function to send this datframe to inlflux
upload_to_influxdb(
    data_frame_to_save=quant_all_first_hr,
    measurement_name=quant_hour_table,
    index_field="timestamp",
    client_name=clients_dict["quantaq"],
    tag_cols_list=["sn", "geo.lat", "geo.lon"],
)

##### 1.3.2 Updating QuantAQ - Hour

Resample the updated data at 1 hour frequency using the minute downloaded data and send it to influxdb

In [93]:
quant_update_hr = resample_quantaq(quantaq_update_min)

#Sometimes the datatypes get mixed up, hence ensure that the 
#data types in this new dataset are consitent with the original one 
quant_update_hr_db = quant_update_hr.astype(quant_all_first_hr.dtypes)

In [94]:
#Use the upload function to send this datframe to inlflux
upload_to_influxdb(data_frame_to_save=quant_update_hr_db, 
                   measurement_name=quant_hour_table, 
                   index_field='timestamp', 
                   client_name=clients_dict["quantaq"],
                   tag_cols_list=['sn','geo.lat','geo.lon'])

Done! Check your influxdb to confirm!


### 1.4 Collocation Quantaq

#### 1.4.1 Minute frequency

First Batch Minute

In [95]:
%%skip

quantaq_all_data_df_coll = get_quantaq_data_devices(devices_sn_list=['MOD-X-00958'], 
                                               start_time='2026-04-16',
                                               end_time='2026-04-17')

quantaq_all_data_df_coll.to_pickle('data-from-monitors/quantaq_all_devices_first_coll.pkl')

In [96]:
quant_all_first_coll = pd.read_pickle('data-from-monitors/quantaq_all_devices_first_coll.pkl')

In [97]:
%%skip
# Use the upload function to send this datframe to inlflux
upload_to_influxdb(
    data_frame_to_save=quant_all_first_coll,
    measurement_name=quant_min_colloc_table,
    index_field="timestamp",
    client_name=clients_dict["quantaq"],
    tag_cols_list=["sn", "geo.lat", "geo.lon"],
)

Update Minute

In [98]:
query_date = """SELECT *
                FROM  'collocation-min-quantaq' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the queery on the influxdb database
table = clients_dict['quantaq'].query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from clarity to influxdb
most_recent_quantaq_df = table.to_pandas()
latest_date_quant = str(most_recent_quantaq_df.time[0])[:10]

#Use two days from now as the end date - for the purposes of buffering
end_date_quant = f'{date.today() + timedelta(days=2)}'

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_quant}')
print(f'start: {end_date_quant}')

start: 2026-06-15
start: 2026-06-18


In [99]:
#%%skip
quantaq_update_min_coll = get_quantaq_data_devices(devices_sn_list=['MOD-X-00958'],
                                             start_time=latest_date_quant,
                                             end_time=end_date_quant)



In [100]:
upload_to_influxdb(data_frame_to_save = quantaq_update_min_coll, 
                   measurement_name=quant_min_colloc_table, 
                   index_field='timestamp', 
                   client_name=clients_dict["quantaq"],
                   tag_cols_list=['sn','geo.lat','geo.lon'])

Done! Check your influxdb to confirm!


#### 1.4.2 Hour Frequency

First Batch Upload

In [101]:
upload_to_influxdb(
    data_frame_to_save=resample_quantaq(quant_all_first_coll),
    measurement_name=quant_hour_colloc_table,
    index_field="timestamp",
    client_name=clients_dict["quantaq"],
    tag_cols_list=["sn", "geo.lat", "geo.lon"],
)

Done! Check your influxdb to confirm!


Update Hour Frequency data

In [102]:
upload_to_influxdb(
    data_frame_to_save=resample_quantaq(quantaq_update_min_coll),
    measurement_name=quant_hour_colloc_table,
    index_field="timestamp",
    client_name=clients_dict["quantaq"],
    tag_cols_list=["sn", "geo.lat", "geo.lon"],
)

Done! Check your influxdb to confirm!


## 2. Clarity
The general guidelines for accessing clarity data through the api can be found [here](https://api-guide.clarity.io/); These guideline use examples based on POSTMAN - a webtool for obtaining data using APIs. Clarity has a python library for retrieving data found [here](https://github.com/a2gov/clarityio). Here we use the python library.

### 2.1 Clarity setup and Automation

First install the clarity library

In [7]:
%%skip
pip install clarityio

Make initial connections with the clarity library


In [8]:
import clarityio # Offers methods to connect to the clarity api
from io import StringIO  # Helps to read file as csv and transform to df
api_connection = clarityio.ClarityAPIConnection(api_key=CLARITY_APIKEY, org=CLARITY_ORGID)

Automate the data retrival process from the clarity website

In [9]:
def request_and_fetch_a_report(start_time, 
                               end_time, 
                               frequency_clarity='minute', 
                               metric_select='default'
                               ):

    """
    A function that retrieves a one-minute-frequency 
    report in csv-wide format for all devices from clarity 
    A maximum of 30 reports can be retrieved in 24 hours

    Args:
        start_time(string) : Starting time for the period of the report 
                             the format of "yyyy-mm-ddThh:mm:ss.sssZ"
        end_time(string)   : End time for the period of the report in 
                             the format of "yyyy-mm-ddThh:mm:ss.sssZ"
        frequency_clarity(string) : The frequency of aggregation due to resampling
                             example: minute(17 mins), hour, day
        metric_select(string)   : The number of metrics to return
                                  "default" or "all". See clarity documentation
                                  for other options
        alldevices (bool)   : Selection for wether to retrieve Data from all devices,
                            if set to False, device_ids must have atleast one element
        device_ids (list)   : a list of devices to retrieve data from

    Return:
        clarity_report_df(DataFrame) : A dataframe report with the clarity 
                                        data in the specified timeframe
    """

    #Define the base part of the clarity url to be used completed in subsequent requests 
    clarity_api_base_url = 'https://clarity-data-api.clarity.io'

    headers = {"x-api-key": CLARITY_APIKEY}

    # request the report
    result = requests.post(url=clarity_api_base_url + "/v2/report-requests",
                           headers=headers,
                           json={
                               "org": CLARITY_ORGID,
                               "report": "datasource-measurements",
                               "allDatasources": True,
                               "outputFrequency": frequency_clarity,
                               "startTime": start_time,
                               "endTime": end_time,
                               "metricSelect": metric_select
                              
                           })
    result.raise_for_status()
    result_json = result.json()
    reportId = result_json['reportId']

    # poll for its completion
    for i in range(12):
        print("sleeping 15 seconds")
        sleep(15)
        print("fetching report status ... ", end="")
        statusUrl = clarity_api_base_url + f"/v2/report-requests/{reportId}"
        result = requests.get(url=statusUrl, headers=headers)
        result.raise_for_status()
        result_json = result.json()
        print(result_json.get("reportStatus"))
        if result_json.get("reportStatus") != 'in-progress':
            break

    print(result_json)

    #Define a variable to store the csv file
    filename = None

    if result_json.get("reportStatus") == 'succeeded':
        # if it succeeded, fetch the resulting files
        for i, url in enumerate(result_json['urls']):
            with requests.get(url=url, stream=True) as result:
                result.raise_for_status()
                filename = f"extract_{i}.csv"
                # stream to disk
                with open(f"{filename}", "w") as f:
                    for chunk in result.iter_content(1024 * 1024, decode_unicode=True):
                        f.write(chunk)
    
    #Remove all null columns to reduce redudancy
    report_df = pd.read_csv(filename)    
    clarity_report_df = report_df.dropna(axis='columns', how="all")

    #Remove the 'status' columns and those with unconventianal units 
    strings_to_drop = ['Num','status','DwerAqi',
                       'UkDefraAqi', 'PhDenrAqi']
    pattern_clarity = '|'.join(strings_to_drop)
    cols_to_drop = clarity_report_df.filter(regex=pattern_clarity, 
                                            axis=1).columns.to_list()
    clarity_report_less_cols = clarity_report_df.copy()
    clarity_report_less_cols.drop(columns=cols_to_drop, axis=1, inplace=True)

    return clarity_report_less_cols

### 2.2 Clarity Minute Data

##### 2.2.1 The First Batch of the Clarity Minute data

Use the request_and_fetch_a_report function to get the intial batch report from clarity and then uploadd it to inlfuxdb.

In [52]:
%%skip

clarity_first_min = request_and_fetch_a_report(start_time="2025-11-05T00:00:00.000Z",
                                                    end_time="2026-03-20T00:00:00.000Z",
                                                    frequency_clarity='minute')


clarity_first_min.to_pickle('data-from-monitors/clarity_first_min.pkl')

In [53]:
clarity_first_min_db = pd.read_pickle('data-from-monitors/clarity_first_min.pkl')

In [54]:
%%skip
#Upload the report to influxdb
upload_to_influxdb(data_frame_to_save=clarity_first_min_db,
                   measurement_name=clarity_min_table,
                   index_field='time',
                   client_name=clients_dict['clarity'],
                   tag_cols_list=['datasourceId','sourceId','sourceType',
                                  'locationLatitude','locationLongitude']
                   )

##### 2.2.2 Updating the Clarity Minute data

To update the clarity database with the latest; first querry for the latest date in the database and use it as the start date to get data from clarity.

In [55]:
query_date = """SELECT *
                FROM  'pilot-min-clarity' 
                ORDER BY time DESC LIMIT 1
             """

# Use the queery on the influxdb database
table = clients_dict['clarity'].query(query=query_date, language="sql")

# Obtain the date to use as a start date in the uplaoding
# the most recent data from clarity to influxdb
most_recent_clarity_df = table.to_pandas()
latest_date_clarity = str(most_recent_clarity_df.time[0])[:10] + "T00:00:00.000Z"

# Use two days from now as the end date - for the purposes of buffering
end_date_clarity = f"{date.today() + timedelta(days=2)}" + "T00:00:00.000Z"

# verify that the date is in the right format and makes sense
print(f"start: {latest_date_clarity}")
print(f"start: {end_date_clarity}")

start: 2026-06-15T00:00:00.000Z
start: 2026-06-18T00:00:00.000Z


Get an updated dataframe from quantaq to upload in the influxdb database

In [56]:
#%%skip
clarity_update_min = request_and_fetch_a_report(start_time=latest_date_clarity,
                                                    end_time=end_date_clarity)


sleeping 15 seconds
fetching report status ... in-progress
sleeping 15 seconds
fetching report status ... succeeded
{'reportId': 'JB7NCAADNY', 'reportStatus': 'succeeded', 'message': 'Ready', 'report': 'datasource-measurements', 'urls': ['https://combined-measurements-export-prd.s3.amazonaws.com/historical/JB7NCAADNY/JB7NCAADNY.csv.gz?AWSAccessKeyId=ASIAQCFEJKFTSAAIUOMP&Signature=1KUEpRkTpGoxqS%2BfCO3KAZsWDzw%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEKz%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIQDH3nvfy4fuoKA%2BrqtBB0ZHrmmm7KYA9CsL7lQWMRd3hgIgU6LZ61nX3C4069lwc46%2BWsnu4ISXrG8zsZwVxw4iPUkqwAQIdRAFGgwwMDQ2Mzk1MTA4ODciDCoquBFVnHNgTn%2BAaCqdBIhv6TMxauAMjcSd%2BuscThP%2F8V3KQJJUkkc9iiG%2B%2FZKyT3NbVChS2xYZhgt06ezcgY6TQGyw4RScWf9lZmidDTfbjdocCRY9hcwSugykxuep68O2SQkBO%2BzAYfnxF8hc7LPN6GseaBPNgsPvtA62i9adQKrckpgDKdSFv5T%2FrzcZJKemnm8x65ucIhmbByNiXtD1V8UVYkYwaGLWwShgF1HwZiG1ZmrAiD15ICokr2BeYEFbXSDhfuGsh7%2F4hgNesqNCCb8TjNQtrSOdPLladLVCT6%2BPuRX9Vak0oR2%2BblWA2oVixOaneDSNPPMg8CiYErCa8tx

In [58]:
# %%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=clarity_update_min,
    measurement_name=clarity_min_table,
    index_field="time",
    client_name=clients_dict["clarity"],
    tag_cols_list=[
        "datasourceId",
        "sourceId",
        "sourceType",
        "locationLatitude",
        "locationLongitude",
    ],
)

Done! Check your influxdb to confirm!


In [59]:
clarity_update_min

,datasourceId,sourceId,sourceType,outputFrequency,time,locationLatitude,locationLongitude,no2ConcIndividual.raw,o3ConcIndividual.raw,pm10ConcMassIndividual.raw,pm1ConcMassIndividual.raw,pm2_5ConcMassIndividual.raw,pm2_5ConcMassIndividual.value,relHumidInternalIndividual.raw,relHumidInternalIndividual.value,temperatureInternalIndividual.raw,temperatureInternalIndividual.value
0,DBFEH9689,A636FW6Q,CLARITY_NODE,minute,2026-06-15T00:00:47.668Z,39.988801,-75.207297,11.05,55.09,6.23,4.00,5.61,6.54,40.23,40.23,30.47,30.47
1,DUWLG7270,A43H6HKC,CLARITY_NODE,minute,2026-06-15T00:01:36.506Z,40.315580,-76.895845,30.33,59.28,11.46,6.77,10.54,6.84,70.65,70.65,25.30,25.30
2,DUPVN4322,A93NM49L,CLARITY_NODE,minute,2026-06-15T00:02:13.343Z,40.247008,-76.846991,33.92,16.04,12.69,8.54,11.92,7.75,68.13,68.13,25.48,25.48
3,DBFEH9689,A636FW6Q,CLARITY_NODE,minute,2026-06-15T00:03:40.292Z,39.988801,-75.207297,11.92,48.05,5.39,3.31,5.08,6.26,40.63,40.63,30.46,30.46
4,DUWLG7270,A43H6HKC,CLARITY_NODE,minute,2026-06-15T00:04:33.567Z,40.315580,-76.895845,23.32,47.58,10.54,6.77,10.31,6.42,71.97,71.97,25.10,25.10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2437,DUPVN4322,A93NM49L,CLARITY_NODE,minute,2026-06-16T10:58:56.491Z,40.247008,-76.846991,17.51,3.52,3.92,2.00,3.69,3.63,73.06,73.06,13.20,13.20
2438,DUWLG7270,A43H6HKC,CLARITY_NODE,minute,2026-06-16T10:59:46.228Z,40.315580,-76.895845,9.10,12.80,2.46,1.54,2.46,3.04,73.88,73.88,14.27,14.27
2439,DBFEH9689,A636FW6Q,CLARITY_NODE,minute,2026-06-16T11:00:02.424Z,39.988801,-75.207297,11.66,9.86,4.46,2.77,4.46,4.80,57.17,57.17,17.51,17.51
2440,DSCFB3568,A47HWW6V,CLARITY_NODE,minute,2026-06-16T11:01:38.276Z,40.264301,-76.851220,14.33,-8.27,5.61,2.23,4.07,4.45,66.14,66.14,15.21,15.21


### 2.3 Clarity Hour Data

##### 2.3.1 First Batch of Clarity

Fetch the data from the clarity website; remember to specify the frequency and select the right metrics

In [19]:
%%skip
clarity_first_hour = request_and_fetch_a_report(
    start_time="2025-11-05T00:00:00.000Z",
    end_time="2026-03-20T00:00:00.000Z",
    frequency_clarity="hour",
    metric_select="all",
)


clarity_first_hour.to_pickle("data-from-monitors/clarity_first_hour")

In [20]:

clarity_first_hour_db = pd.read_pickle('data-from-monitors/clarity_first_hour')

In [21]:
%%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=clarity_first_hour_db,
    measurement_name=clarity_hour_table,
    index_field="startOfPeriod",
    client_name=clients_dict['clarity'],
    tag_cols_list=[
        "datasourceId",
        "sourceId",
        "sourceType",
        "locationLatitude",
        "locationLongitude",
    ],
)

##### 2.3.2 Updating Clarity Hour Data to influxdb

In [16]:
query_date = """SELECT *
                FROM  'pilot-hour-clarity' 
                ORDER BY time DESC LIMIT 1
             """

# Use the queery on the influxdb database
table = clients_dict["clarity"].query(query=query_date, language="sql")

# Obtain the date to use as a start date in the uplaoding
# the most recent data from clarity to influxdb
most_recent_clarity_df = table.to_pandas()
latest_date_clarity = str(most_recent_clarity_df.time[0])[:10] + "T00:00:00.000Z"

# Use two days from now as the end date - for the purposes of buffering
end_date_clarity = f"{date.today() + timedelta(days=2)}" + "T00:00:00.000Z"

# verify that the date is in the right format and makes sense
print(f"start: {latest_date_clarity}")
print(f"start: {end_date_clarity}")

start: 2026-06-02T00:00:00.000Z
start: 2026-06-18T00:00:00.000Z


In [17]:
clarity_update_df_hour = request_and_fetch_a_report(
    start_time=latest_date_clarity,
    end_time=end_date_clarity,
    frequency_clarity="hour",
    metric_select="all",
)

sleeping 15 seconds
fetching report status ... in-progress
sleeping 15 seconds
fetching report status ... succeeded
{'reportId': 'JB1RND8KVK', 'reportStatus': 'succeeded', 'message': 'Ready', 'report': 'datasource-measurements', 'urls': ['https://combined-measurements-export-prd.s3.amazonaws.com/historical/JB1RND8KVK/JB1RND8KVK.csv.gz?AWSAccessKeyId=ASIAQCFEJKFTTTEFBGHG&Signature=DmaclxeHYYMUV%2F4IYPPxsFiuGx0%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEKv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJGMEQCIHlnwOBfuSdZrWsymwvT4ztGzLfzp7yczNZXXONi7sIrAiAwoSBIG6SQgMABD%2FmgEnD5jrS2wWUHVEbza3zDpcbm1irABAh0EAUaDDAwNDYzOTUxMDg4NyIMctFr1W%2BSayg7FmtkKp0EyYjnZT%2Bwwqva3GUuSto%2FyG971sXmbP6y72rjKhJyzoVSTZSU5fGwMoIInjWI17PYvHC66nWlRDXJ39O9Z9nAihYspLewzcMECGfdKWDLWOs1mdVpQPiRdaZIZFRWVbQl7lLlESQzWGW0t5sSaam1r9zjSG%2FGO1ezvwfTwBgMU0Y7trlwY2zxCmFWaWBwn%2Brs0O5Ozahm9LexZ5sHhP01MeWG%2BQIA0GXmP4KiID3H1%2FJOMfy8AYWESvN4XJeqlE9tRbw605tyUHTcA%2BS5EXDoB8JpsOZkthsjIWjUzIF2ZXg8nm%2BQiGUZ1inwSMIpg1CqYBBhJY7B3P8

Sometimes clarity drops columns as it updates its databases. To ensure consistency add the dropped columns to the new dataframe

In [ ]:
# First identify the columns that exist in the first batch table but not in the update
dropped_cols = list(
    clarity_first_hour_db.columns[
        ~clarity_first_hour_db.columns.isin(clarity_update_df_hour.columns)
    ]
)

# Add the columns and assign them to zeros
clarity_update_df_hour[dropped_cols] = 0

In [23]:
# Sometimes the datatypes get mixed up, hence ensure that the
# data types in this new dataset are consistent with the original one
clarity_update_hour_db = None
try:
    clarity_update_hour_db = clarity_update_df_hour.astype(
        clarity_first_hour_db.dtypes, errors="raise"
    )
except:
    # Define the columns that should be turned to integers
    col_int = list(
        clarity_first_hour_db.columns[clarity_first_hour_db.dtypes == "int64"]
    )

    #Fill the unknowns with 0 to prevent data conversion errors
    clarity_update_df_hour[col_int] = clarity_update_df_hour[col_int].fillna(0)

    #Then convert those columns to their respective datatypes
    clarity_update_df_hour[col_int] = clarity_update_df_hour[col_int].astype("float")
    clarity_update_hour_db = clarity_update_df_hour
    

To ensure further consistence in the dataset; let's obtain a sample dataset from influxdb and use it to verify datasets

In [ ]:
%%skip
query_sample_table = """SELECT *
                FROM  'pilot-hour-clarity' 
                ORDER BY time 
             """

# Use the queery on the influxdb database
table_sample_clarity = clients_dict["clarity"].query(query=query_sample_table, language="sql")
table_sample_df = table_sample_clarity.to_pandas()

#Save the table locally to prevent future downloads
table_sample_df.to_pickle(
    "data-from-monitors/clarity-hour-nov25-june26-influx-download.pkl"
)


Save the above locally so as not to downlaod it again in future 

In [ ]:
clarity_hour_sample_from_influx = pd.read_pickle(
    "data-from-monitors/clarity-hour-nov25-june26-influx-download.pkl"
)

In [73]:
# We use the above data from influxdb to ensure that the datatypes match
clarity_sample_without_time = clarity_hour_sample_from_influx.loc[
    :, clarity_hour_sample_from_influx.columns != "time"
]


In [76]:
clarity_update_hour_db.loc[:, clarity_update_hour_db.columns != "startOfPeriod"] = (
    clarity_update_hour_db.loc[:, clarity_update_hour_db.columns != "startOfPeriod"].astype(
        clarity_sample_without_time.dtypes, errors="ignore"
    )
)

Do a last bit of conversion of the datypes that didn't convert 

In [78]:
cols_to_convert_to_float = ['no2Conc1HourMean.value','no2Conc1HourMeanUsEpaAqi.value']
clarity_update_hour_db[cols_to_convert_to_float] = clarity_update_hour_db[cols_to_convert_to_float].astype('float') 

In [79]:
cols_to_convert_to_int = ['pm2_5ConcMassNowcastUsEpaAqi.value', 'pm2_5ConcMassNowcastUsEpaAqi.raw','pm10ConcMassNowcastUsEpaAqi.raw']
clarity_update_hour_db[cols_to_convert_to_int] = clarity_update_hour_db[cols_to_convert_to_int].astype('int64') 

In [80]:
upload_to_influxdb(data_frame_to_save=clarity_update_hour_db,
                   measurement_name=clarity_hour_table,
                   index_field='startOfPeriod',
                   client_name=clients_dict['clarity'],
                   tag_cols_list=['datasourceId','sourceId','sourceType',
                                  'locationLatitude','locationLongitude']
                   )

Done! Check your influxdb to confirm!


### 2.4 Collocation Clarity

In [ ]:
#Define the devices to collocate
clarity_coll_devices = ["DUPVN4322"]

#### 2.4.1 Minute Clarity

In [ ]:
%%skip

clarity_first_min_coll = request_and_fetch_a_report(start_time="2026-04-16T00:00:00.000Z",
                                                    end_time="2026-04-17T00:00:00.000Z",
                                                    frequency_clarity='minute'                                      
                                                    
                                                    )

clarity_first_min_coll.to_pickle('data-from-monitors/clarity_first_min_coll.pkl')

Filter for only those to be used in collocation

In [ ]:
clarity_first_min_coll = pd.read_pickle('data-from-monitors/clarity_first_min_coll.pkl')
clarity_first_min_coll_db = clarity_first_min_coll[
    clarity_first_min_coll["datasourceId"].isin(clarity_coll_devices)
]

In [ ]:
%%skip
#Upload the report to influxdb
upload_to_influxdb(data_frame_to_save=clarity_first_min_coll_db,
                   measurement_name=clarity_min_colloc_table,
                   index_field='time',
                   client_name=clients_dict['clarity'],
                   tag_cols_list=['datasourceId','sourceId','sourceType',
                                  'locationLatitude','locationLongitude']
                   )

In [ ]:
query_date = """SELECT *
                FROM  'collocation-min-clarity' 
                ORDER BY time DESC LIMIT 1
             """

# Use the queery on the influxdb database
table = clients_dict['clarity'].query(query=query_date, language="sql")

# Obtain the date to use as a start date in the uplaoding
# the most recent data from clarity to influxdb
most_recent_clarity_df = table.to_pandas()
latest_date_clarity = str(most_recent_clarity_df.time[0])[:10] + "T00:00:00.000Z"

# Use two days from now as the end date - for the purposes of buffering
end_date_clarity = f"{date.today() + timedelta(days=2)}" + "T00:00:00.000Z"

# verify that the date is in the right format and makes sense
print(f"start: {latest_date_clarity}")
print(f"start: {end_date_clarity}")

start: 2026-05-25T00:00:00.000Z
start: 2026-05-28T00:00:00.000Z


In [ ]:
#%%skip
clarity_update_min_coll = request_and_fetch_a_report(start_time=latest_date_clarity,
                                                    end_time=end_date_clarity)
clarity_update_min_coll_db = clarity_update_min_coll[
    clarity_update_min_coll["datasourceId"].isin(clarity_coll_devices)
]

sleeping 15 seconds
fetching report status ... succeeded
{'reportId': 'JB2G3GQ2JV', 'reportStatus': 'succeeded', 'message': 'Ready', 'report': 'datasource-measurements', 'urls': ['https://combined-measurements-export-prd.s3.amazonaws.com/historical/JB2G3GQ2JV/JB2G3GQ2JV.csv.gz?AWSAccessKeyId=ASIAQCFEJKFTY3MV3L4D&Signature=hNQf6M37kHqpOm7IThZMHZEY11A%3D&x-amz-security-token=IQoJb3JpZ2luX2VjELr%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQD8IPwn4GOONIMIJH7K9NqGdfnYFnS3ND9m6blaSa5TIAIhAP1m8GRB6JGtpqc%2B0Wr2bq7ZuRFCsXDbIYlKcG3cWg9MKskECIP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBRoMMDA0NjM5NTEwODg3IgzAl5d4Pgqr4NycIqAqnQTQIZTKWzCb%2FgC8nT4MekpvYp6wNR5bdk7wHL%2F%2FbBSRJVdeF0dvFX8dSc0VnwBv%2FljpBfY8KKqygtCGgU9DJTn9odVITioMBCFWRfmo1LR7fZuQJk%2BWaKWUBbYQl2A4SJhQS%2F0%2Fl1ZJJKonacfTU%2B57CVjUflEeu8cXg%2FZxh9SVL8%2FI8g8fE8WKouQFLl8x4uwJdM6X1yTj7%2BDuWdlLei%2BQFqUd4esnJsqCS%2Bj3Q4cTIC8UWg96izC1bcBlhEPSPazA3x%2FnUgVfBKdZdElhRiYgN7RsJM%2FX8WCP9rDOihva5%2B9lNiwjau3sRY%2BJWaYYc8gBXTd6X%2B4K3e90UJ81oa

In [ ]:
# %%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=clarity_update_min_coll_db,
    measurement_name=clarity_min_colloc_table,
    index_field="time",
    client_name=clients_dict["clarity"],
    tag_cols_list=[
        "datasourceId",
        "sourceId",
        "sourceType",
        "locationLatitude",
        "locationLongitude",
    ],
)

Done! Check your influxdb to confirm!


#### 2.4.2 Clarity Hour collocation

In [ ]:
%%skip

clarity_first_hour_coll = request_and_fetch_a_report(start_time="2026-04-16T00:00:00.000Z",
                                                    end_time="2026-04-17T00:00:00.000Z",
                                                    frequency_clarity='hour',
                                                    metric_select='all'                                      
                                                    
                                                    )

clarity_first_hour_coll.to_pickle('data-from-monitors/clarity_first_hour_coll.pkl')

In [ ]:
clarity_first_hour_coll = pd.read_pickle('data-from-monitors/clarity_first_hour_coll.pkl')
clarity_first_hour_coll_db = clarity_first_hour_coll[
    clarity_first_hour_coll["datasourceId"].isin(clarity_coll_devices)
]

In [ ]:
%%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=clarity_first_hour_coll_db,
    measurement_name=clarity_hour_colloc_table,
    index_field="startOfPeriod",
    client_name=clients_dict['clarity'],
    tag_cols_list=[
        "datasourceId",
        "sourceId",
        "sourceType",
        "locationLatitude",
        "locationLongitude",
    ],
)

In [ ]:
query_date = """SELECT *
                FROM  'collocation-hour-clarity' 
                ORDER BY time DESC LIMIT 1
             """

# Use the queery on the influxdb database
table = clients_dict['clarity'].query(query=query_date, language="sql")

# Obtain the date to use as a start date in the uplaoding
# the most recent data from clarity to influxdb
most_recent_clarity_df = table.to_pandas()
latest_date_clarity = str(most_recent_clarity_df.time[0])[:10] + "T00:00:00.000Z"

# Use two days from now as the end date - for the purposes of buffering
end_date_clarity = f"{date.today() + timedelta(days=2)}" + "T00:00:00.000Z"

# verify that the date is in the right format and makes sense
print(f"start: {latest_date_clarity}")
print(f"start: {end_date_clarity}")

start: 2026-05-25T00:00:00.000Z
start: 2026-05-28T00:00:00.000Z


In [ ]:
#%%skip
clarity_update_hour_coll = request_and_fetch_a_report(start_time=latest_date_clarity,
                                                    end_time=end_date_clarity)
clarity_update_hour_coll_db = clarity_update_hour_coll[
    clarity_update_hour_coll["datasourceId"].isin(clarity_coll_devices)
]

sleeping 15 seconds
fetching report status ... succeeded
{'reportId': 'JBBGHQDZ8N', 'reportStatus': 'succeeded', 'message': 'Ready', 'report': 'datasource-measurements', 'urls': ['https://combined-measurements-export-prd.s3.amazonaws.com/historical/JBBGHQDZ8N/JBBGHQDZ8N.csv.gz?AWSAccessKeyId=ASIAQCFEJKFTY3MV3L4D&Signature=hsc4VysWwTtmFAm3wEtXLw0ri4c%3D&x-amz-security-token=IQoJb3JpZ2luX2VjELr%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQD8IPwn4GOONIMIJH7K9NqGdfnYFnS3ND9m6blaSa5TIAIhAP1m8GRB6JGtpqc%2B0Wr2bq7ZuRFCsXDbIYlKcG3cWg9MKskECIP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBRoMMDA0NjM5NTEwODg3IgzAl5d4Pgqr4NycIqAqnQTQIZTKWzCb%2FgC8nT4MekpvYp6wNR5bdk7wHL%2F%2FbBSRJVdeF0dvFX8dSc0VnwBv%2FljpBfY8KKqygtCGgU9DJTn9odVITioMBCFWRfmo1LR7fZuQJk%2BWaKWUBbYQl2A4SJhQS%2F0%2Fl1ZJJKonacfTU%2B57CVjUflEeu8cXg%2FZxh9SVL8%2FI8g8fE8WKouQFLl8x4uwJdM6X1yTj7%2BDuWdlLei%2BQFqUd4esnJsqCS%2Bj3Q4cTIC8UWg96izC1bcBlhEPSPazA3x%2FnUgVfBKdZdElhRiYgN7RsJM%2FX8WCP9rDOihva5%2B9lNiwjau3sRY%2BJWaYYc8gBXTd6X%2B4K3e90UJ81oa

In [ ]:
# %%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=clarity_update_hour_coll_db,
    measurement_name=clarity_hour_colloc_table,
    index_field="time",
    client_name=clients_dict["clarity"],
    tag_cols_list=[
        "datasourceId",
        "sourceId",
        "sourceType",
        "locationLatitude",
        "locationLongitude",
    ],
)

Done! Check your influxdb to confirm!


## 3. AirGradient

The general instructions on how to access Airgradient data by API can be found [here](https://www.airgradient.com/professional/cities/undp-toolkit/operating/airgradient-api/). Specific instructions on different options avialable can be found [here](https://api.airgradient.com/public/docs/api/v1/)

### 3.1 AirGradient Automation and setup

Since Airgradient is limitted to a maximum of 10 days in retrieving data, create a function that extend this to more days. 
This is helpful especially in retirving the initial batch of data.

In [103]:
def get_airgradient_data_one(
    location_id,
    start_date,
    end_date,
):
    """
    A function that retrived 5-minute data
    from airgradient in a given time range

    Args:
        location_id (string) : The location id whose data is to be retrieved
        start_date (string)   : A starting date in the format of yyyymmdd
                                Data starts at 00:00:00 utc on this date
        end_date(string)      : An end date in the formart of yyyymmdd
                                Data ends at 00:00:00 utc on this date

    Returns:
        airgradient_df (dataframe) : A dataframe with the airgradient data
    """

    # First convert all the dates to date formart so as to easily perform date operations
    start_date = datetime.strptime(start_date, "%Y%m%d")
    end_date = datetime.strptime(end_date, "%Y%m%d")

    # Initialize the dataframe that will be used to collect other dataframe
    airgradient_df_list = list()

    # Since airgradient allows to retrieve only a maximum of 10 data records
    # We use a loop to segment the data into digestable intervals
    while start_date < end_date:       

        interval_start_date = start_date
        interval_end_date = interval_start_date + timedelta(days=8)

        # Make the new end date to be the starting date for the next cyle
        if interval_end_date > end_date:
            interval_end_date = end_date

        # Convert the dates abve in an airgradient format and
        # use them in the api calls to get the data
        date_from = f"{str(interval_start_date).replace('-','')[:8]}T000000Z"
        date_to = f"{str(interval_end_date).replace('-','')[:8]}T000000Z"

        end_point_params = {
            "location_id": location_id,
            "token": AIRGRADIENT_APIKEY,
            "from": date_from,
            "to": date_to,
        }

        # Define the intial text to be used in the api call
        initial_url = f"https://api.airgradient.com/public/api/v1/locations/{location_id}/measures/past?"
        json_file = requests.get(initial_url, params=end_point_params)

        # Reformat the resulting json file for easy conversion into a df
        json_reformat = json.loads(json_file.content.decode("utf-8"))
        dataframe = pd.json_normalize(json_reformat)

        # Make a list of dataframes to be used to
        # make one single dataframe to be returned
        airgradient_df_list.append(dataframe)

        # Set the start date once again to allow
        # increaments and to break the loop
        start_date = interval_end_date

    # Convert the collected dataframes into one single dataframe
    airgradient_df = pd.concat(airgradient_df_list)

    
    return airgradient_df

Given that we shall be trying to download the data for different locations in a given time frame at once, let's write a function that can take multiple locations.

In [104]:
def get_airgradient_data_many(locations_ids, 
                              start_date,
                                end_date):
    """ 
    Takes a list of location ids (strings) and returns a single combined dataframe
    with all the 5-minute interval data collected in a given time range
    """

    #First setup a list to collect the dataframes 
    airgradient_list_many = []

    #collect them in to the list 
    for location in locations_ids:
        airgradient_list_many.append(get_airgradient_data_one(start_date=start_date,
                                                               end_date=end_date,
                                                               location_id=location))

    #Produce one final combined dataframe with all the dataframes 
    airgradient__many_locations_df = pd.concat(airgradient_list_many)  

    #Ensure that the time column is in date format 
    # for easy intergation in influxdb

    airgradient__many_locations_df['timestamp'] = (pd.to_datetime
                                                   (airgradient__many_locations_df
                                                    ['timestamp']))
    
    #Remove any empty rows or columns     
    airgradient__many_locations_df.dropna(axis=1, 
                                          how='all',
                                          inplace=True)
    
    # To avoid, exceeding downloading limits, allow some
    print("sleeping 20 seconds")
    sleep(20)

    return airgradient__many_locations_df                                              



Airgradeint data comes with a number of unwanted fields. This redudancy can sometimes result in compatibility issues especially in updating the database with new data. Create a function that removed some of the unwanted columns 

In [105]:
def clean_airgradient_data(airgradient_raw_df):

    """" 
    Takes an airgradeint dataframe with raw data and 
    ensures that only a few wanted columns are retained 

    """

    #First identify the unawated columns to drop
    cols_to_drop_ag = ['wifi','model', 'tvocIndex', 'noxIndex', 'datapoints']

    airgadient_clean = airgradient_raw_df.drop(columns=cols_to_drop_ag)
    
    return airgadient_clean

Create a resampling function to aggregate the parameters per hour. 
Airgradient does not offer a direct way to pull resampled data for more than 7 days from the resampling date.

In [106]:
def resample_airgradient(df, frequency='1h'):
    
    """
    The resample_airgradient takes a quantaq dataframe and 
    Returns a dataframe resampled at a given frequency

    Args:
        df (object)        :A dataframe of airgradient data 
                            collected at minute level
        frequency (string) :Frequency of sampling/aggregating
                            see the resample function in pandas
    
    Return:
        df_all_sampled (object) :A dataframe with data resampled 
                             at the stated interval
    """

    #First convert the timestamp to datetime to pervent any errors
    #df['timestamp'] = pd.to_datetime(df['timestamp'])

    #Define the dictionary to be used in resampling so that we can 
    # get the mean of numerical fields wiothout losing categorical ones

    cols_dict = {'pm01': 'mean', 'pm02':'mean', 
             'pm02_corrected':'mean', 'pm10_corrected':'mean', 
             'pm10':'mean', 'pm01_corrected':'mean', 
             'pm003Count':'mean', 'atmp':'mean', 'rhum':'mean',
             'rco2':'mean','atmp_corrected':'mean', 'rhum':'mean',
             'rhum_corrected':'mean', 'rco2_corrected':'mean', 
             'locationId':'first',
             'locationName':'first',
             'serialno':'first'             
             }
    
    #Define a dataframe to combine all the dataframes for each device
    df_all_sampled = []    

    for device_sn in pd.unique(df.serialno):
        #get a dataframe for each device separately
        df_device = df[df['serialno']==device_sn]

        #Resample the dataframe of the device 
        df_device_sampled = (df_device.resample('1h',
                                               on='timestamp',).
                                               agg(cols_dict).
                                               reset_index())

        #Add it to the overall dataframe to be returned 
        df_all_sampled.append(df_device_sampled)

    df_all_sampled = pd.concat(df_all_sampled, ignore_index=True)

    #Remove the values without locations 
    df_all_sampled.dropna(axis=0, subset=['locationId'], inplace=True)
     

    return df_all_sampled  

### 3.2 AirGradient Minute Data

Aigradient data is collected at a minimum of 5-minute interval

##### 3.2.1 First Batch of AirGradient Minute

In [107]:
%%skip
# State the locations for which data is to be collected
# For now, the locations are Harrisburg-001, Harrisburg-003-02, 
# Collocation-Harrisburg-005,  Harrisburg-006, 
locations_airgradient = ["175884", "182974", "183437", "183438"]

# Collect the data from the allocations in
# a time of apparoximately a month to date
airgradient_first_min = get_airgradient_data_many(
    start_date="20251001", end_date="20260324", locations_ids=locations_airgradient
)

airgradient_first_min.to_pickle("data-from-monitors/airgradient_first_min.pkl")

Store the data as pickle for easy accessibility and clean it using the clean_airgradient function to ensure that only the useful columns remail

In [108]:
#%%skip
airgradient_min_first_pickle = pd.read_pickle("data-from-monitors/airgradient_first_min.pkl")
airgradient_min_first_db = clean_airgradient_data(airgradient_min_first_pickle)

Upload the data to influxdb

In [109]:
%%skip
#Upload the report to influxdb
upload_to_influxdb(data_frame_to_save=airgradient_min_first_db,
                   measurement_name=airgradient_min_table,
                   index_field='timestamp',
                   client_name=clients_dict['airgradient'],
                   tag_cols_list=['locationId','locationName',
                                  'serialno']
                   )

Ensure that the influxdb field data types match their original data types collected directed from the monitors

In [110]:
query_params = """ DESCRIBE 'pilot-min-airgradient' """
params = clients_dict['airgradient'].query(query=query_params, language='sql')
params.to_pandas()

,column_name,data_type,is_nullable
0,atmp,Float64,YES
1,atmp_corrected,Float64,YES
2,batteryVoltage,Utf8,YES
3,locationId,"Dictionary(Int32, Utf8)",YES
4,locationName,"Dictionary(Int32, Utf8)",YES
5,panelVoltage,Utf8,YES
6,pm003Count,Int64,YES
7,pm01,Float64,YES
8,pm01_corrected,Float64,YES
9,pm02,Float64,YES


##### 3.2.2 Updating AirGradient Minute

In [111]:
query_date = """SELECT *
                FROM  'pilot-min-airgradient' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the queery on the influxdb database
table = clients_dict['airgradient'].query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from airgradient to influxdb
most_recent_airgradient_df = table.to_pandas()
latest_date_airgradient = str(most_recent_airgradient_df.
                              time[0])[:10].replace('-','')

#Use two days from now as the end date - for the purposes of buffering
end_date_airgradient = f'{date.today() + timedelta(days=2)}'[:10].replace('-','')

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_airgradient }')
print(f'end: {end_date_airgradient}')

start: 20260615
end: 20260618


In [112]:
#Get the updated data 

# For now, the locations are Harrisburg-001, Harrisburg-003-02, 
# Collocation-Harrisburg-005,  Harrisburg-006, Harrisburg-007, 
# Harrisburg-008, Harrisburg-009, Harribsurg-010, Harribsurg-011 
# Harribsurg-012, Harribsurg-013 
locations_airgradient = ["175884" ,"182974", 
                         "183437", "183438", "190626", 
                         "190624", "191464", "192213", 
                         "192214", "192212", "192211"]

#Use the dates and the locations to get the airgradient data for updates
airgradient_min_update = get_airgradient_data_many(start_date=latest_date_airgradient,
                                                         end_date=end_date_airgradient,
                                                         locations_ids=locations_airgradient)




sleeping 20 seconds


Clean the updates downlaoded and ensure that the data types match those in the first batch

In [113]:
# Clean up and save the updated data
airgradient_min_update_db = clean_airgradient_data(airgradient_min_update)

#Fill the "na" values with zeros to prevent data conversion errors
airgradient_min_update_db = airgradient_min_update_db.fillna(0)

# Ensure that the data types are matching
airgradient_min_update_db = airgradient_min_update_db.astype(
    airgradient_min_first_db.dtypes,
)

Now send the updates to inlfuxdb

In [114]:
upload_to_influxdb(
    data_frame_to_save=airgradient_min_update_db,
    measurement_name=airgradient_min_table,
    index_field="timestamp",
    client_name=clients_dict["airgradient"],
    tag_cols_list=["locationId", "locationName", "serialno"],
)

Done! Check your influxdb to confirm!


### 3.3 Airgradient Hour Data

##### 3.3.1 First Batch AirGardient - Hour

In [115]:
#%%skip
# Resample on the minute data to form 1 hour bins
ag_resample_hour = resample_airgradient(airgradient_min_first_db, frequency="hour")

# Ensure that the data types are matching
ag_resample_hour = ag_resample_hour.astype(airgradient_min_first_db.dtypes)

In [116]:
#%%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=ag_resample_hour,
    measurement_name=airgradient_hour_table,
    index_field="timestamp",
    client_name=clients_dict['airgradient'],
    tag_cols_list=["locationId", "locationName", "serialno"],
)

Done! Check your influxdb to confirm!


##### 3.3.2 Update AirGradient Hour

In [117]:
# Resample on the minute data to form 1 hour bins
ag_resample_hour_update = resample_airgradient(airgradient_min_update_db, frequency="hour")

# Ensure that the data types are matching
ag_resample_hour_update = ag_resample_hour_update.astype(airgradient_min_first_db.dtypes)

In [118]:
# %%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=ag_resample_hour_update,
    measurement_name=airgradient_hour_table,
    index_field="timestamp",
    client_name=clients_dict["airgradient"],
    tag_cols_list=["locationId", "locationName", "serialno"],
)

Done! Check your influxdb to confirm!


In [119]:
query_params = """ DESCRIBE 'pilot-hour-airgradient' """
params = clients_dict['airgradient'].query(query=query_params, language='sql')
pd.DataFrame(params.to_pandas())

,column_name,data_type,is_nullable
0,atmp,Float64,YES
1,atmp_corrected,Float64,YES
2,locationId,"Dictionary(Int32, Utf8)",YES
3,locationName,"Dictionary(Int32, Utf8)",YES
4,pm003Count,Int64,YES
5,pm01,Float64,YES
6,pm01_corrected,Float64,YES
7,pm02,Float64,YES
8,pm02_corrected,Float64,YES
9,pm10,Float64,YES


### 3. 4 Airgradient Gases
As we we wait for airgradient to officially allow access to the gases dataset through API, we can download the data manually and simply upload it to influxdb

##### 3.4.1 Automation for Airgradient gases data

In [120]:
def clean_airgradient_gases_data(airgradient_gases_raw_df):
    """ "
    Takes an airgradeint dataframe with raw data and
    ensures that only a few wanted columns are retained

    """
    # Rename the columns measure0 to measure5 to
    # their respective actual scientific names

    actual_names_dict = {
        "measure0": "O3 Working Electrode (mV)",
        "measure1": "O3 Auxiliary Electrode (mV)",
        "measure2": "NO2 Working Electrode (mV)",
        "measure3": "NO2 Auxiliary Electrode (mv)",
        "measure5": "O3 ppb",
        "measure6": "NO2 ppb",
    }

    airgradient_gases_raw_df = airgradient_gases_raw_df.rename(
        columns=actual_names_dict
    )

    # First identify the unwanted columns to drop
    cols_to_drop_ag = [
        "TVOC (ppb)",
        "TVOC index",
        "NOX index",
        "Location Type",
        "Location Group",
    ]
    airgadient_gases_clean = airgradient_gases_raw_df.drop(columns=cols_to_drop_ag)

    # Remove other irrelevant columns named measure...
    airgadient_gases_clean = airgadient_gases_clean.loc[
        :, ~airgadient_gases_clean.columns.str.contains("easure")
    ]

    # Ensure that the dates attributes are of the right domain (time)
    airgadient_gases_clean[["Local Date/Time", "UTC Date/Time"]] = (
        airgadient_gases_clean[["Local Date/Time", "UTC Date/Time"]].apply(
            pd.to_datetime
        )
    )

    # Ensure that the columns are well named
    airgadient_gases_clean = airgadient_gases_clean.clean_names(
        remove_special=True, strip_underscores=True
    )

    return airgadient_gases_clean

Create a function to be used to combine the dataframes once they are downloaded from the different devices 

In [121]:
def combine_ag_gases_dataframes(df_list):
    """

    Returns one dataframe with a data from all the devices with gas measurements
    """

    # Instatiate the dataframe collector
    ag_gases_combined_df = pd.concat(df_list)

    return ag_gases_combined_df

##### 3.4.2 Download, Clean and Upload the data

Get the latest date of the data in the database to help in determiniogn the downloading range

In [122]:
query_date = """SELECT *
                FROM  'collocation-hour-gases-ag-clean' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the queery on the influxdb database
table = clients_dict['airgradient'].query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from airgradient to influxdb
most_recent_airgradient_df = table.to_pandas()
latest_date_airgradient = str(most_recent_airgradient_df.
                              time[0])[:10].replace('-','')

#Use two days from now as the end date - for the purposes of buffering
end_date_airgradient = f'{date.today() + timedelta(days=2)}'[:10].replace('-','')

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_airgradient }')
print(f'end: {end_date_airgradient}')

start: 20260601
end: 20260618


In [123]:
airgradient_hour_gases_table


'collocation-hour-gases-ag-clean'

Upload the downloaded for cleaning

In [124]:
# First Obtain the data from devices that measure gases
hbg_10_df = pd.read_csv(
    "data-from-monitors/airgradient-gases-collocation/Collocation-Harrisburg-010.csv"
)
new_11_df = pd.read_csv(
    "data-from-monitors/airgradient-gases-collocation/Collocation-Philly-New-011.csv"
)
mon_12_df = pd.read_csv(
    "data-from-monitors/airgradient-gases-collocation/Collocation-Philly-Montgomery-012.csv"
)

Do some basic cleaning of the data and combine the dataframes

In [125]:
ag_gases_df = combine_ag_gases_dataframes(
    [
        clean_airgradient_gases_data(hbg_10_df),
        clean_airgradient_gases_data(new_11_df),
        clean_airgradient_gases_data(mon_12_df),
    ]
)

Upload the data to influxdb

In [126]:
# %%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=ag_gases_df,
    measurement_name=airgradient_hour_gases_table,
    index_field="utc_date_time",
    client_name=clients_dict["airgradient"],
    tag_cols_list=["location_id", "location_name", "sensor_id"],
)

Done! Check your influxdb to confirm!


## 4. EPA Data
We need API data to give us an average bachmark of what to expect from our models. 
By using data from locations close to our low cost devices, we can get a range of expectation of expectation. 
More guidelines on how to retrieve EPA data using API can be found [here](https://aqs.epa.gov/aqsweb/documents/data_api.html/).

### 4.1 EPA Setup and lists
EPA has a set of code lists for states, counties and most parameters. 
It is important to get a summary of these codes.

In [31]:
%%skip
#First register the the email address to receive a key 
signup_url = 'https://aqs.epa.gov/data/api/signup?email=izabayoalain@gmail.com'
response_json = requests.get(signup_url)
json.loads(response_json.content)

In [32]:
#get some import lists and tables that will be crucial in retreiving data
states_url = 'https://aqs.epa.gov/data/api/list/states?'
states_params = {'email':EPA_ID,'key':EPA_KEY}
states_json = requests.get(states_url, states_params)
states_epa = pd.json_normalize(json.loads(states_json.content)['Data'])
states_epa.to_pickle("states_codes")


In [33]:
#Pennsylvania Counties codes 
counties_url = 'https://aqs.epa.gov/data/api/list/countiesByState?'
counties_params = {'email':EPA_ID,'key':EPA_KEY, 'state':42}
counties_json = requests.get(counties_url, counties_params)
counties_epa = pd.json_normalize(json.loads(counties_json.content)['Data'])
#counties_epa.to_pickle("counties_codes")

In [34]:
counties_epa

,code,value_represented
0,001,Adams
1,003,Allegheny
2,005,Armstrong
3,007,Beaver
4,009,Bedford
...,...,...
62,125,Washington
63,127,Wayne
64,129,Westmoreland
65,131,Wyoming


Dauphin - 043, Franklin - 055, Lancaster - 071, Cumberland - 041, Adams(Arendtsville) - 001, Philadelphia - 101

In [35]:
#%%skip
#Pennsylvania, Duaphin site codes 
sites_url = 'https://aqs.epa.gov/data/api/list/sitesByCounty?'
sites_params = {'email':EPA_ID,'key':EPA_KEY, 'state':42, 'county':'001'}
sites_json = requests.get(sites_url , sites_params)
sites_dauphin_epa = pd.json_normalize(json.loads(sites_json.content)['Data'])
#sites_dauphin_epa.to_pickle("sites_dauphin_epa")

In [36]:
sites_dauphin_epa

,code,value_represented
0,0001,Arendtsville
1,0002,Biglerville
2,8001,None
3,9000,None
4,9991,Arendtsville


From the above, we derive the the site codes for the stations in Harrisburg and Hershey to be; 0401 - Harrisburg, 1100 - Hershey, 0001 - Arendtsville

In [37]:
%%skip
#Parameters 
class_url = 'https://aqs.epa.gov/data/api/list/classes?'
class_params = {'email':EPA_ID,'key':EPA_KEY}
class_json = requests.get(class_url, class_params)
classes_dauphin_epa = pd.json_normalize(json.loads(class_json.content)['Data'])
classes_dauphin_epa.to_pickle("classes_dauphin_epa")

In [38]:
#%%skip
#Check for all parameter codes 
param_url = 'https://aqs.epa.gov/data/api/list/parametersByClass?'
param_params = {'email':EPA_ID,'key':EPA_KEY, 'pc':'CRITERIA'}
param_json = requests.get(param_url, param_params)
param_dauphin_epa = pd.json_normalize(json.loads(param_json.content)['Data'])
#param_dauphin_epa.to_pickle("param_dauphin_epa")

In [39]:
param_dauphin_epa

,code,value_represented
0,14129,Lead (TSP) LC
1,42101,Carbon monoxide
2,42401,Sulfur dioxide
3,42602,Nitrogen dioxide (NO2)
4,44201,Ozone
5,81102,PM10 Total 0-10um STP
6,85129,Lead PM10 LC FRM/FEM
7,88101,PM2.5 - Local Conditions


From the above we notice the code values of some of the major pollutants we shall monitor. 42101 - Carbon monoxide, 44201	- Ozone, 88101 - PM2.5 - Local Conditions, 42401 - Sulphur dioxide, 42602 - Nitogen dioxide(NO2), 81102 - PM10 Total

Let's get the data now!

In [40]:
sample_url = 'https://aqs.epa.gov/data/api/sampleData/bySite?'
sample_params = {'email':EPA_ID,'key':EPA_KEY, 
                 'param':42401,                  
                 'bdate':20251216, 
                 'edate':20251218,
                 'state':42, 'county':'001', 
                 'site':'0001'}
sample_json = requests.get(sample_url, sample_params)
sample_sites_data_02_2026 = pd.json_normalize(json.loads(sample_json.content)['Data'])
#sample_sites_data_02_2026.to_pickle("sample_sites_data_02_2026")

### 4.2 AirNow Data
Since EPA provides data that is too old - sometimes the latest data maybe 6 months old; we can try to access AirNow data instead. More information about accessing airnow data can be found [here](https://docs.airnowapi.org/)

##### 4.2.1 Automation and Setup

In [7]:
def get_airnow_data(start_date, 
                    end_date, 
                    geo_box_list = ["-77.378871,39.705655,-76.192348,40.549861"]):
    """
    A function that returns OZONE,PM25,PM10,CO,NO2 and SO2 data from the
     EPA monitors - if the monitors have them

    Args:
        start_date(string) : date in the format of yyy-mm-ddThh
        end_date (string)  : date in the format of yyy-mm-ddThh
                            The total duration querried should 
                            be less than approximately 2 months
        geo_box_list (list)   : a list of strings demarcating 
                            our areas of interest. Each string
                            is made up of four points (rectangle vertices)                      

    Returns:
        all_geo_df (dataframe) : A datframe with OZONE,PM25,
                                        PM10,CO,NO2 and SO2 data
    """

    #Instatiate a dataframe to collect the dataframes of different geographies
    geo_df_list = []

    # Let's add each geo_box to the api url and get the data for each location
    for vertices_string in geo_box_list:
        # First define the argumnets to be used in the api call
        url_airnow_1 = f"https://www.airnowapi.org/aq/data/?"
        url_airnow_date = f"startDate={start_date}&endDate={end_date}"
        url_airnow_param = f"&parameters=OZONE,PM25,PM10,CO,NO2,SO2"
        url_airnow_geog = f"&BBOX={vertices_string}"
        url_airnow_format = f"&dataType=B&format=application/json&verbose=1"
        url_airnow_monitor = f"&monitorType=0&includerawconcentrations=0"
        url_airnow_key = f"&API_KEY={AIRNOW_KEY}"

        airnow_url = (
            url_airnow_1
            + url_airnow_date
            + url_airnow_param
            + url_airnow_geog
            + url_airnow_format
            + url_airnow_monitor
            + url_airnow_key
        )
        # print(airnow_url)
        # Use the above to get the jason files
        airnow_json = requests.get(airnow_url).content
        # Ensure that the json file is decoded for easy manipulation
        airnow_json_decoded = airnow_json.decode("utf-8")
        # Obtain a dataframe
        airnow_df = pd.json_normalize(json.loads(airnow_json_decoded))

        #add the dataframe to the collector 
        geo_df_list.append(airnow_df)
        all_geo_df = pd.concat(geo_df_list, axis='index')

    

    return all_geo_df

In [8]:
def clean_airnow_data(airnow_raw_df):
    """ 
    Returns a clean version of the generated dataframe;
    It renames the time field and drops fields that wont
    be helpful during data analysis 

    """
    # Ensure that the timestamp is properly renamed and converted accodingly
    airnow_raw_df["UTC"] = pd.to_datetime(airnow_raw_df["UTC"])
    airnow_raw_time = airnow_raw_df.rename({"UTC": "timestamp"}, axis=1)

    #Drop columns that will most likley not be useful
    cols_to_drop = ['AgencyName']
    airnow_raw_clean = airnow_raw_time.drop(columns= cols_to_drop)

    #Remove missed values which are at times reported as -999.0
    airnow_raw_clean_v1 = airnow_raw_clean[airnow_raw_clean['Value']>0]

    #Ensure that category is an int
    cols_to_int = ['Category', 'AQI'] 
    airnow_raw_clean_v1[cols_to_int] = airnow_raw_clean_v1[cols_to_int].astype("int")

    return airnow_raw_clean_v1


##### 4.3.2 First Batch from Airnow

Use the automated functions to get and clean the data. The data should be for nearly all the EPA stations in central PA

In [9]:
# %%skip
# Define the argument parameters in retriving the airnow data
harriburg_geo_box = "-77.378871,39.705655,-76.192348,40.549861"
philly_geo_box = "-75.664725,39.735784,-74.824271,40.265981"
geo_list = [harriburg_geo_box, philly_geo_box]

# Use the get function to obtain the data from the api
airnow_first = get_airnow_data(
    start_date="2026-04-01T23", end_date="2026-04-30T23", geo_box_list=geo_list
)

In [10]:
#Clean the data to ensure it is consistent and less redudant 
airnow_first_clean = clean_airnow_data(airnow_first)
airnow_first_clean.to_pickle("data-from-monitors/airnow_first_clean.pkl")

C:\Users\Owner\AppData\Local\Temp\ipykernel_19528\2427810947.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airnow_raw_clean_v1[cols_to_int] = airnow_raw_clean_v1[cols_to_int].astype("int")


In [11]:
airnow_first_clean_db = pd.read_pickle('data-from-monitors/airnow_first_clean.pkl')

Upload the new data in influxdb. Ensure that all the fields that are descriptive are stated under the tags_cols_list

In [12]:
#%%skip
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=airnow_first_clean_db,
    measurement_name=airnow_hour_table,
    index_field="timestamp",
    client_name=clients_dict['zz-airnow'],
    tag_cols_list=[
        "Latitude",
        "Longitude",
        "SiteName",
        "Parameter",
        "FullAQSCode",
        "IntlAQSCode",
        "Unit"
    ]
)

Done! Check your influxdb to confirm!


##### 4.2.3 Update Airnow

First obtain the new start and end time

In [13]:
query_date = """SELECT *
                FROM  'pilot-hour-airnow' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the querry on the influxdb database
table = clients_dict['zz-airnow'].query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from clarity to influxdb
most_recent_airnow = table.to_pandas()
latest_date_airnow= str(most_recent_airnow.time[0])[:10] + 'T00'

#Use two days from now as the end date - for the purposes of buffering
end_date_airnow = f'{date.today() + timedelta(days=2)}' + 'T00'

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_airnow}')
print(f'end: {end_date_airnow}')

start: 2026-06-18T00
end: 2026-06-20T00


In [14]:
# Define the argument parameters in retriving the airnow data
harriburg_geo_box = "-77.378871,39.705655,-76.192348,40.549861"
philly_geo_box = "-75.664725,39.735784,-74.824271,40.265981"
geo_list = [harriburg_geo_box, philly_geo_box]

# Get the data now
airnow_update = get_airnow_data(
    start_date=latest_date_airnow, end_date=end_date_airnow, geo_box_list=geo_list
)
airnow_update_db = clean_airnow_data(airnow_update)

C:\Users\Owner\AppData\Local\Temp\ipykernel_19528\2427810947.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airnow_raw_clean_v1[cols_to_int] = airnow_raw_clean_v1[cols_to_int].astype("int")


In [15]:
# Upload the report to influxdb
upload_to_influxdb(
    data_frame_to_save=airnow_update_db,
    measurement_name=airnow_hour_table,
    index_field="timestamp",
    client_name=clients_dict['zz-airnow'],
    tag_cols_list=[
        "Latitude",
        "Longitude",
        "SiteName",
        "Parameter",
        "FullAQSCode",
        "IntlAQSCode",
        "Unit"
    ]
)

Done! Check your influxdb to confirm!


## 5. Collocation data cleaninng

We obtain the collocation data for all the monitors used within a the collocation time interval, identify the key parameters to include in the report and then produce the report that will include the r-sqaured, root-mean-square errors and the means. The parameters to invistigate are PM2.5, NO2, O3, CO


### 5.1 Data download from sensors

#### 5.1.1 QuantAQ Data download

The Harribsurg quantaq Monitor

In [165]:
query_quantaq_hbg_02 = """SELECT "time", "geo.lat", "geo.lon", "sn", "pm25", "no2","o3","co"
                FROM  'pilot-hour-quantaq' 
                WHERE  time >= timestamp '2026-04-16T00:00:00.000Z' AND time <= timestamp '2026-06-16T00:00:00.000Z' AND 
                        "geo.lat" = '40.247'
             """

# Use the queery on the influxdb database
table_quantaq_hbg_02 = clients_dict["quantaq"].query(query=query_quantaq_hbg_02, language="sql")
table_quantaq_hbg_02_df = table_quantaq_hbg_02.to_pandas()


The Philly Quantaq monitor 

In [166]:
query_quantaq_philly_01 = """SELECT "time", "geo.lat", "geo.lon", "sn", "pm25", "no2","o3","co"
                FROM  'pilot-hour-quantaq' 
                WHERE  time >= timestamp '2026-05-02T00:00:00.000Z' AND time <= timestamp '2026-06-07T00:00:00.000Z' AND 
                        "geo.lat" = '39.992'
             """

# Use the queery on the influxdb database
table_quantaq_philly_01 = clients_dict["quantaq"].query(query=query_quantaq_philly_01, language="sql")
table_quantaq_philly_01_df = table_quantaq_philly_01.to_pandas()

In [167]:
def clean_quantaq_collocation(
    quant_df1,
    quant_df2,
    lat_location={"39.992": "Collocation-Philly-New-002", "40.247": "collocation-hbg-001"},
):
    """
    Returns combined dataframes from two quantaq sources with new standardized,
    renamed and potentially additional columns

    quant_df1(dataframe) : A dataframe with sensor data from quantaq
    quant_df2(dataframe) ; a dataframe with sensor data from quantaq
    lat_location (dict) : key-value pair of the latitude and 
                            and a description of the location and device
                            to appear in the table

    """

    # First combine the two columns
    combined_quantaq_df = pd.concat([quant_df1, quant_df2], axis=0)

    # Create a dictionary with updated names of the columns
    quantaq_new_names_dict = {
        "geo.lat": "latitude",
        "geo.lon": "longitude",
        "sn": "device_sn",
    }
    combined_quantaq_df = combined_quantaq_df.rename(columns=quantaq_new_names_dict)

    # Introduce a column for manufacturer/supplier
    combined_quantaq_df["manufacturer"] = "quantaq"

    # Add the location/site name column and the device-description
    for lat_key in list(lat_location.keys()):
        combined_quantaq_df.loc[
            combined_quantaq_df["latitude"] == lat_key, "location_device_details"
        ] = lat_location[lat_key]
        

    # Reorder the columns for simplicity
    combined_quantaq_df = combined_quantaq_df[
        [
            "time",
            "latitude",
            "longitude",
            "device_sn",
            "manufacturer",
            "location_device_details",
            "pm25",
            "no2",
            "o3",
            "co",
        ]
    ]

    return combined_quantaq_df

Let's use the above function to combine the philly and Harrisburg data from quantaq

In [168]:
quantaq_philly_hbg = clean_quantaq_collocation(
    table_quantaq_hbg_02_df, table_quantaq_philly_01_df
)

#### 5.1.2 Clarity data download 

The Harrisburg site clarity

In [169]:
query_clarity_hbg_01 =  """SELECT "time", "locationLatitude", "locationLongitude",
                                 "sourceId","pm2_5ConcMass1HourMean.value",
                                 "o3Conc1HourMean.raw","no2Conc1HourMean.raw","coConc1HourMean.raw"

                           FROM    'pilot-hour-clarity' 

                           WHERE   time >= timestamp '2026-04-16T00:00:00.000Z' AND 
                                 time <= timestamp '2026-06-17T00:00:00.000Z' AND 
                                 "sourceId" = 'A93NM49L' AND 
                                 "locationLatitude" = '40.24700813'
                        """

# Use the queery on the influxdb database
table_clarity_hbg_01 = clients_dict["clarity"].query(
    query=query_clarity_hbg_01, language="sql"
)
table_clarity_hbg_01_df = table_clarity_hbg_01.to_pandas()

Ideally, when relocation is done properly, followed by timely update of the clarity system - we can querry for a given location using the datasourceId. However in the case of our collocation, we did not do proper record keeping properly hence we have to use a combination of the latitude and the device sn to querry for the location 

The Philly-North East West (NEW) site 

In [170]:
query_clarity_philly_02 =  """SELECT "time", "locationLatitude", "locationLongitude",
                                 "sourceId","pm2_5ConcMass1HourMean.value",
                                 "o3Conc1HourMean.raw","no2Conc1HourMean.raw","coConc1HourMean.raw"

                           FROM    'pilot-hour-clarity' 

                           WHERE   time >= timestamp '2026-05-01T00:00:00.000Z' AND 
                                 time <= timestamp '2026-06-07T00:00:00.000Z' AND 
                                 "sourceId" = 'A44MFTF3' AND 
                                 "locationLatitude" = '39.99195745'
                        """

# Use the queery on the influxdb database
table_clarity_philly_02 = clients_dict["clarity"].query(
    query=query_clarity_philly_02, language="sql"
)
table_clarity_philly_02_df = table_clarity_philly_02.to_pandas()

The Philly - Montgomery (MON) site

In [171]:
query_clarity_mon_04 =  """SELECT "time", "locationLatitude", "locationLongitude",
                                 "sourceId","pm2_5ConcMass1HourMean.value",
                                 "o3Conc1HourMean.raw","no2Conc1HourMean.raw","coConc1HourMean.raw"

                           FROM    'pilot-hour-clarity' 

                           WHERE   time >= timestamp '2026-05-01T00:00:00.000Z' AND 
                                 time <= timestamp '2026-06-07T00:00:00.000Z' AND 
                                 "sourceId" = 'A636FW6Q' AND 
                                 "locationLatitude" = '39.98880106'
                        """

# Use the queery on the influxdb database
table_clarity_mon_04 = clients_dict["clarity"].query(
    query=query_clarity_mon_04, language="sql"
)
table_clarity_mon_04_df = table_clarity_mon_04.to_pandas()

Let's clean and combine the data from all the clarity sensors 

In [172]:
def clean_sensor_collocaton(dataframes_list, 
                             new_column_names_dict, 
                             location_device_details_dict, 
                             manufacturer):
    
    """ 
    Returns a single dataframe with all the data from clarity sensors,
    with columns properly renamed 

    Args:
        dataframes_list (list) : A list of dataframes with data 
                                from sensors from different sites
        location_device_details (dict) : key-value pair of the latitude(key) and 
                                    and a description of the location and device
                                    to appear in the table (value)
        new_column_names_dict : A key-value pair of old names and 
                                new names of the columns
        manufacturer (string) : Name of manufacturer eg clarity, quantaq etc.

    """

    # First combine the two columns

    combined_sensors_df = pd.concat(dataframes_list, axis=0)

    # Create a dictionary with updated names of the columns
    combined_sensors_df = combined_sensors_df.rename(columns=new_column_names_dict)

    # Introduce a column for manufacturer/supplier
    combined_sensors_df["manufacturer"] = manufacturer

    # Add the location/site name column and the device-description
    for lat_key in list(location_device_details_dict.keys()):
        combined_sensors_df.loc[
            combined_sensors_df["latitude"] == lat_key, "location_device_details"
        ] = location_device_details_dict[lat_key]

    # Reorder the columns for simplicity
    combined_sensors_df = combined_sensors_df[
        [
            "time",
            "latitude",
            "longitude",
            "device_sn",
            "manufacturer",
            "location_device_details",
            "pm25",
            "no2",
            "o3",
            "co",
        ]
    ]

    return combined_sensors_df

Let's combine the data from the clarity sensors

In [173]:
# Define the required arguments
clarity_df_list = [
    table_clarity_hbg_01_df,
    table_clarity_philly_02_df,
    table_clarity_mon_04_df,
]
clarity_new_cols = {
    "locationLatitude": "latitude",
    "locationLongitude": "longitude",
    "sourceId": "device_sn",
    "pm2_5ConcMass1HourMean.value": "pm25",
    "o3Conc1HourMean.raw": "o3",
    "no2Conc1HourMean.raw": "no2",
    "coConc1HourMean.raw": "co",
}
clarity_loc_details = {
    "40.24700813": "collocation-hbg-1 (DUPVN4322)",
    "39.99195745": "Collocation-Philly-New-002 (DGYDF0388)",
    "39.98880106": "Collocation-Philly-Montgomery-004 (DBFEH9689)",
}
manuf = "clarity"

# Let's use the function to combine and clean the columns 
clarity_philly_hbg_df = clean_sensor_collocaton(
    dataframes_list=clarity_df_list,
    new_column_names_dict=clarity_new_cols,
    location_device_details_dict=clarity_loc_details,
    manufacturer=manuf,
)

#### 5.1.3 Air Gradient data download

Air gradient that is stored in influxdb does not contain the ozone and no2 data since during collocation airgradient had not enabled accessing ozone and no2 data by api.
Hence in this analysis, we downloaded the airgradeint data manually and then we shall analyse the csv files accordingly.


In [174]:
# Find find all the csv files
airgradient_csv_files = glob.glob(
    "data-from-monitors/airgradient-collocation-data-june26/*.csv"
)

# Combine th files into one single datframe
airgradient_collocation_df = pd.concat(
    (pd.read_csv(airgradient_csv) for airgradient_csv in airgradient_csv_files),
    ignore_index=True,
)

We can first retain only the columns we shall need and then refine them to meet the quantaq and clarity cleaned columns. In retaining the columns we keep measure 5 and 6 too based on Aigradient guidlines below;  

measure0: O3 Working Electrode (mV)  
measure1: O3 Auxiliary Electrode (mV)  
measure2: NO2 Working Electrode (mV)  
measure3: NO2 Auxiliary Electrode (mv)  
measure5: O3 ppb  
measure6: NO2 ppb


In [175]:
# Define the columns to retain
ag_cols_to_retain = [
    "Location ID",
    "Location Name",
    "Sensor ID",
    "UTC Date/Time",
    "PM2.5 (μg/m³) corrected",
    "measure5",
    "measure6",
]

# Retain the columns
airgradient_collocation_filtered = airgradient_collocation_df[ag_cols_to_retain]

# Ensure that the time column is in time-date format
airgradient_collocation_filtered["UTC Date/Time"] = pd.to_datetime(
    airgradient_collocation_filtered["UTC Date/Time"], errors="coerce"
)

# Update the location name for Collocation-Philly-New-011
airgradient_collocation_filtered.loc[
    (airgradient_collocation_filtered["Location ID"] == 192214), "Location Name"
] = "Collocation-Philly-New-011"

C:\Users\Owner\AppData\Local\Temp\ipykernel_19528\3205052787.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  airgradient_collocation_filtered["UTC Date/Time"] = pd.to_datetime(


In [176]:
def clean_airgradient_collocation(
    airgradient_combined_df, new_ag_cols_dict, location_lats_longs_dict
):
    """
    Returns a new dataframe with all airgradient sensor data
    with columns matching those of other devices

    args:
        airgradient_combined_df: A dataframe with all the airgradient
                                data from different sensors
        new_colnames_dict : A key-value with keys as former names
                            and values as the new column names
        location_lats_longs_dict : A key-value pair with keys as location-ids and
                                    a vlaue of lats and longs as elements of a list

    """

    # First rename the columns
    airgradient_combined_df = airgradient_combined_df.rename(columns=new_ag_cols_dict)

    # Add the the lat and longs columns and values
    for location_id in list(location_lats_longs_dict.keys()):
        # Create the latitude column
        airgradient_combined_df.loc[
            (airgradient_combined_df["Location ID"] == location_id), "latitude"
        ] = location_lats_longs_dict[location_id][0]

        # Create the longitude column
        airgradient_combined_df.loc[
            (airgradient_combined_df["Location ID"] == location_id), "longitude"
        ] = location_lats_longs_dict[location_id][-1]

    # Add the manufacturer column for ideintification when data is mixed with other sensors
    airgradient_combined_df["manufacturer"] = "airgradient"

    # Add a co column for uniformity with other dataframes(clarity and aq)
    airgradient_combined_df["co"] = pd.NA

    # Drop the location id as it is no longer needed
    airgradient_combined_final_df = airgradient_combined_df.drop(
        columns=["Location ID"]
    )

    # Reoder the columns accordingly
    airgradient_combined_final_df = airgradient_combined_final_df[
        [
            "time",
            "latitude",
            "longitude",
            "device_sn",
            "manufacturer",
            "location_device_details",
            "pm25",
            "no2",
            "o3",
            "co",
        ]
    ]

    return airgradient_combined_final_df

In [177]:
# Define the arguments to use
new_aigrad_cols_dict = {
    "Location Name": "location_device_details",
    "Sensor ID": "device_sn",
    "UTC Date/Time": "time",
    "PM2.5 (μg/m³) corrected": "pm25",
    "measure5": "o3",
    "measure6": "no2",
}

location_lats_longs_ag = {
    183437: ["40.24700813", "-76.84699099"],
    192730: ["39.99195745", "-75.08114984"],
    190626: ["39.98880106", "-75.20729697"],
    190624: ["39.99195745", "-75.08114984"],
    192213: ["40.24700813", "-76.84699099"],
    192214: ["39.99195745", "-75.08114984"],
    192212: ["39.98880106", "-75.20729697"],
}

# Now clean and update the airgadeint dataset
airgradient_hbg_philly_df = clean_airgradient_collocation(
    airgradient_combined_df=airgradient_collocation_filtered,
    new_ag_cols_dict=new_aigrad_cols_dict,
    location_lats_longs_dict=location_lats_longs_ag,
)

### 5.2 Combine and clean the datasets from different manufacturers

In [178]:
# Since all the columns have the same names;
#  we can try to directly concatenate the data
hbg_philly_collocation_sensors = pd.concat(
    [quantaq_philly_hbg, clarity_philly_hbg_df, airgradient_hbg_philly_df], axis=0
).reset_index(drop=True)

C:\Users\Owner\AppData\Local\Temp\ipykernel_19528\853789287.py:3: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  hbg_philly_collocation_sensors = pd.concat(


In [179]:
# Test to confirm that no data was lost in the concatenation process
hbg_philly_collocation_sensors.shape[0] == (
    quantaq_philly_hbg.shape[0]
    + clarity_philly_hbg_df.shape[0]
    + airgradient_hbg_philly_df.shape[0]
)

True

We can now do some basic cleaning by for example fomarting the time column, and potentially combining all the parameters into one column 

In [180]:
#First ensure that the time column is a time date-format
#uTC is set to true to return time localized to utc-zone
hbg_philly_collocation_sensors['time'] = pd.to_datetime(hbg_philly_collocation_sensors['time'], utc=True)

#Indicate in the column name that the time is in utc 
hbg_philly_collocation_sensors = hbg_philly_collocation_sensors.rename(columns={"time":"time_utc"})

The lats and lomngs of the quantaq devices are either truncated on worngly written. We can rectify this here!

In [181]:
# Chnge the lat and longs of the Harrisburg quantaq device
hbg_philly_collocation_sensors.loc[(hbg_philly_collocation_sensors["device_sn"] == "MOD-X-00958"), 
                                   ["latitude", "longitude"]] = ["40.24700813", "-76.84699099"]

# Change the lats and longs of the philly quantaq device 
hbg_philly_collocation_sensors.loc[(hbg_philly_collocation_sensors["device_sn"] == "MOD-X-00959"), 
                                   ["latitude", "longitude"]] = ["39.99195745", "-75.08114984"]


To simplify the querrying process, we pivot the data so that all the measures are in one column

In [182]:
# Define the columns that we won't have to collapse into rows
columns_to_keep = [
    "time_utc",
    "latitude",
    "longitude",
    "device_sn",
    "manufacturer",
    "location_device_details",
]
# Define the columns that we want to collapse into measures
columns_to_collapse = ["pm25", "no2", "o3", "co"]

hbg_philly_collocation_long = pd.melt(
    hbg_philly_collocation_sensors,
    id_vars=columns_to_keep,
    value_vars=columns_to_collapse,
    var_name="parameter",
    value_name="pm_gases_measures",
)

We can change the parameter names to match the names by epa 

In [183]:
#Create a dictionary to change the names 
new_param_names = {'pm25':'PM2.5', 'no2':'NO2', 'o3':'OZONE', 'co':'CO'}

#We now change the records 
hbg_philly_collocation_long['parameter'] = hbg_philly_collocation_long['parameter'].replace(new_param_names)


To match the site names with those by EPA, we change the names and rename the column

In [184]:
# Change the name for the Harribsurg location device
hbg_philly_collocation_long.loc[
    (hbg_philly_collocation_long["latitude"] == "40.24700813"),
    "location_device_details",
] = "Harrisburg"

# The montgomery site
hbg_philly_collocation_long.loc[
    (hbg_philly_collocation_long["latitude"] == "39.99195745"),
    "location_device_details",
] = "Philly-MON"

# The NEW site
hbg_philly_collocation_long.loc[
    (hbg_philly_collocation_long["latitude"] == "39.98880106"),
    "location_device_details",
] = "Philly-NEW"

In [185]:
hbg_philly_collocation_long[hbg_philly_collocation_long['location_device_details']=='Collocation-Philly-New-002-02-New']

,time_utc,latitude,longitude,device_sn,manufacturer,location_device_details,parameter,pm_gases_measures


In [194]:
#Change the column name to match the new naming of sites 
hbg_philly_collocation_long = hbg_philly_collocation_long.rename(columns={'location_device_details':'site_name'})

### 5.3 Airnow collocation data

We First download the data within the collocation range, filter it and then do some basic cleaning to ready it for merging with other datasets from sensors

Next, we filter for only ozone, pm2.5, no2 and co ad for only the Harrisburg, MON and New sites.
We then do some baic cleaning and renaming of the columns to ensure that the airnow data matches the ones from sensor table

In [196]:
query_airnow_collocation = """SELECT "time","Latitude", "Longitude", "Parameter", "Value",
                                   "SiteName", "IntlAQSCode"
                            FROM  'pilot-hour-airnow' 
                            WHERE  time >= timestamp '2026-04-15T00:00:00.000Z' AND 
                                   time <= timestamp '2026-06-16T23:00:00.000Z' AND 
                            "SiteName" IN ('Harrisburg','NEW','MON') AND
                            "Parameter" IN ('OZONE', 'PM2.5', 'NO2', 'CO')
             """

# Use the queery on the influxdb database
table_airnow_collocation = clients_dict['zz-airnow'].query(query=query_airnow_collocation, language="sql")
table_airnow_collocation_df = table_airnow_collocation.to_pandas()

In [197]:
# Rename the columns to match the names of the sensors df
new_cols_airnow = {
    "time": "time_utc",
    "Latitude": "latitude",
    "Longitude": "longitude",
    "Parameter": "parameter",
    "Value": "pm_gases_measures",
    "SiteName": "site_name",
    "IntlAQSCode": "device_sn",
}
table_airnow_renamed = table_airnow_collocation_df.rename(columns=new_cols_airnow)
# Add the manufacturer column
table_airnow_renamed["manufacturer"] = "EPA"

# Modify the lats and longs to match the table we have
# The Harrisburg site
table_airnow_renamed.loc[
    (table_airnow_renamed["device_sn"] == "840420430401"), ["latitude", "longitude"]
] = ["40.24700813", "-76.84699099"]

# The lats and longs for the NEW-Philly site
table_airnow_renamed.loc[
    (table_airnow_renamed["device_sn"] == "840421010048"), ["latitude", "longitude"]
] = ["39.99195745", "-75.08114984"]
# The Montgomery site
table_airnow_renamed.loc[
    (table_airnow_renamed["device_sn"] == "840421010076"), ["latitude", "longitude"]
] = ["39.98880106", "-75.20729697"]

# Update the philly site names in the location_devices-details
# For the new site
table_airnow_renamed.loc[
    (table_airnow_renamed["site_name"] == "NEW"),
    ["site_name"],
] = "Philly-NEW"
# For the Montgomery site
table_airnow_renamed.loc[
    (table_airnow_renamed["site_name"] == "MON"),
    ["site_name"],
] = "Philly-MON"

### 5.4 Collocation data Cleaned  

We combine the data from sensors, check for consistence of datasets and then export it as csv file for analysis 

The time column data type should be ocnverted and the a new column with local time zone should be created. 
The lats and longs should also be converted to floats

In [ ]:
# Combine the datasets
collocation_sensors_airnow_combined = pd.concat(
    [hbg_philly_collocation_long, table_airnow_renamed]
)

# Make the time the right data type
collocation_sensors_airnow_combined["time_utc"] = pd.to_datetime(
    collocation_sensors_airnow_combined["time_utc"], utc=True
)

# Create a new column with local time zone
collocation_sensors_airnow_combined["time_US_eastern"] = (
    collocation_sensors_airnow_combined["time_utc"].dt.tz_convert("America/New_York")
)

# Remove the timzeone offset suffux
collocation_sensors_airnow_combined["time_US_eastern"] = (
    collocation_sensors_airnow_combined["time_US_eastern"].dt.tz_localize(None)
)

# Rmeove the eastern time column so that it can later be added
eastern_time_pop = collocation_sensors_airnow_combined.pop("time_US_eastern")
# Put the new column at the beginning
collocation_sensors_airnow_combined.insert(0, "time_US_eastern", eastern_time_pop)

# Drop the utc time column
collocation_sensors_airnow_combined.drop(columns=["time_utc"], inplace=True)

In [244]:
# Let's ensure that the lats and longs are numeric 
collocation_sensors_airnow_combined[['latitude','longitude']] = collocation_sensors_airnow_combined[['latitude','longitude']].apply(pd.to_numeric)

Save the data for futher analysis

In [ ]:
%%skip
# Save to github foilder
collocation_sensors_airnow_combined.to_csv(
    "data-from-monitors/collocation-report-june26-folder/hbg-philly-collocation-apr-june-2026.csv"
)

# Save to project folder
collocation_sensors_airnow_combined.to_csv(
    "G:/My Drive/General/profession/CI4CJ/mapping_air_quality_har_phil/1-collocation-analysis/data-collocation/datasets/hbg-philly-collocation-apr-june-2026.csv"
)